crypto/rand/rand_lib.c

```diff
@@ -90,15 +91,117 @@ int rand_rdcpu(void)
+ * randomness from the system, and that |min_len| bytes will do.
+ */
+size_t drbg_entropy_from_system(RAND_DRBG *drbg,
+                                unsigned char **pout,
+                                int entropy, size_t min_len, size_t max_len)
+{
+    int i;
+
+
+    if (min_len > (size_t)drbg->size) {
+        /* Should not happen.  See comment near RANDOMNESS_NEEDED. */
+        min_len = drbg->size;
+    }
+
+    if (rand_drbg.filled) {
+        /* Re-use what we have. */
+        *pout = drbg->randomness;
+        return drbg->size;
+    }
+
+    /* If we don't have enough, try to get more. */
+    CRYPTO_THREAD_write_lock(rand_bytes.lock);
+    for (i = RAND_POLL_RETRIES; rand_bytes.curr < min_len && --i >= 0; ) {
+        CRYPTO_THREAD_unlock(rand_bytes.lock);
+        RAND_poll();
+        CRYPTO_THREAD_write_lock(rand_bytes.lock);
+    }
+
+    /* Get desired amount, but no more than we have. */
+    if (min_len > rand_bytes.curr)
+        min_len = rand_bytes.curr;
+    if (min_len != 0) {
+        memcpy(drbg->randomness, rand_bytes.buff, min_len);
+        rand_drbg.filled = 1;
+        /* Update amount left and shift it down. */
+        rand_bytes.curr -= min_len;
+        if (rand_bytes.curr != 0)
+            memmove(rand_bytes.buff, &rand_bytes.buff[min_len], rand_bytes.curr);
+    }
+    CRYPTO_THREAD_unlock(rand_bytes.lock);
+    return min_len;
+}
+
+size_t drbg_entropy_from_parent(RAND_DRBG *drbg,
+                                unsigned char **pout,
+                                int entropy, size_t min_len, size_t max_len)
+{
+    int st;
+
+    if (min_len > (size_t)drbg->size) {
+        /* Should not happen.  See comment near RANDOMNESS_NEEDED. */
+        min_len = drbg->size;
+    }
+
+    /* Get random from parent, include our state as additional input. */
+    st = RAND_DRBG_generate(drbg->parent, drbg->randomness, min_len, 0,
+                            (unsigned char *)drbg, sizeof(*drbg));
+    if (st == 0)
+        return 0;
+    drbg->filled = 1;
+    return min_len;
+}
+
+void drbg_release_entropy(RAND_DRBG *drbg, unsigned char *out)
+{
+    drbg->filled = 0;
+    OPENSSL_cleanse(drbg->randomness, sizeof(drbg->randomness));
+}
+
 DEFINE_RUN_ONCE_STATIC(do_rand_init)
 {
     int ret = 1;
+
 #ifndef OPENSSL_NO_ENGINE
     rand_engine_lock = CRYPTO_THREAD_lock_new();
     ret &= rand_engine_lock != NULL;
 #endif
     rand_meth_lock = CRYPTO_THREAD_lock_new();
     ret &= rand_meth_lock != NULL;
+
+    rand_bytes.lock = CRYPTO_THREAD_lock_new();
+    ret &= rand_bytes.lock != NULL;
+    rand_bytes.curr = 0;
+    rand_bytes.size = MAX_RANDOMNESS_HELD;
+    /* TODO: Should this be secure malloc? */
+    rand_bytes.buff = malloc(rand_bytes.size);
+    ret &= rand_bytes.buff != NULL;
+
+    rand_drbg.lock = CRYPTO_THREAD_lock_new();
+    ret &= rand_drbg.lock != NULL;
+    rand_drbg.size = RANDOMNESS_NEEDED;
+    rand_drbg.randomness = OPENSSL_malloc(rand_drbg.size);
```

Comment:
Ideally, this ought to be secure too.

The per thread ones perhaps but not will be okay in the absence of knowledge of the number of threads.

The per SSL ones not.